src/primitives.rs

```diff
@@ -0,0 +1,133 @@
+use bytes::Buf;
+use std::fmt::{Error, Formatter};
+
+#[derive(FromPrimitive)]
+pub enum ApiKey {
+    Metadata = 3,
+    ApiVersions = 18,
+}
+
+pub enum ApiVersion {
+    Version0,
+    Version1,
+    Version2,
+}
+
+#[derive(Debug)]
+pub enum DecodeError {
+    BufferUnderflow,
+    BufferOverflow,
+    BadData,
+}
+
+#[derive(Debug, PartialEq)]
+pub struct HeaderRequest {
+    pub api_key: i16,
+    pub api_version: i16,
+    pub correlation_id: i32,
+    pub client_id: String,
+}
+
+impl std::fmt::Display for HeaderRequest {
+    fn fmt(&self, _: &mut Formatter) -> Result<(), Error> {
+        println!(
+            "api_key: {}, api_version: {}, correlation_id: {}, client_id: {}",
+            self.api_key, self.api_version, self.correlation_id, self.client_id
+        );
+        Ok(())
+    }
+}
+
+impl FromByte for HeaderRequest {
+    fn decode(buf: &mut Buf) -> Result<Self, DecodeError> {
+        Ok(HeaderRequest {
+            api_key: decode_buffer(buf)?,
+            api_version: decode_buffer(buf)?,
+            correlation_id: decode_buffer(buf)?,
+            client_id: decode_buffer(buf)?,
+        })
+    }
+}
+
+#[derive(Debug)]
+pub struct HeaderResponse {
+    pub correlation_id: i32,
+}
+
+pub trait BodyRequest {}
+pub struct DecodedRequest {
+    pub size: i32,
+    pub header: HeaderRequest,
+    pub body: Box<dyn BodyRequest>,
+}
+
+impl DecodedRequest {
+    pub fn decode(buf: &mut Buf) -> Result<Self, DecodeError> {
+        let size: i32 = decode_buffer(buf)?;
+        let header: HeaderRequest = decode_buffer(buf)?;
+
+        let body: Box<dyn BodyRequest> = match FromPrimitive::from_i16(header.api_key) {
+            Some(ApiKey::ApiVersions) => Box::new(BodyApiVersionRequest::decode(buf)?),
+            Some(ApiKey::Metadata) => Box::new(BodyMeatdataApiVersionRequest::decode(buf)?),
+            _ => return Err(DecodeError::BadData),
+        };
+
+        Ok(DecodedRequest { size, header, body })
+    }
+}
+
+pub trait FromByte: Sized {
+    fn decode(buf: &mut Buf) -> Result<Self, DecodeError>;
+}
+
+impl FromByte for i32 {
+    fn decode(buf: &mut Buf) -> Result<Self, DecodeError> {
+        if buf.remaining() < 4 {
+            //Err(DecodeError::BufferUnderflow);
+        }
+        Ok(buf.get_i32_be())
+    }
+}
+
+impl FromByte for i16 {
+    fn decode(buf: &mut Buf) -> Result<Self, DecodeError> {
+        if buf.remaining() < 4 {
+            //Err(DecodeError::BufferUnderflow);
+        }
+        Ok(buf.get_i16_be())
+    }
+}
+
+impl FromByte for String {
+    fn decode(buf: &mut Buf) -> Result<Self, DecodeError> {
+        let length: i16 = decode_buffer(buf)?;
+        if length == 0 || length == -1 {
+            return Ok(String::from(""));
+        }
+        // Length should not be smaller than -1 according to the Kafka protocol
+        if length < -1 {
+            // ToDo: throw error
+            // return Ok(());
+        }
+        // Any smart way to read a String from buffer?
```

Comment:
The `String` type represents a heap-allocated string, so this is the right approach if you want to decode an owned `String`. Since you don't own the buffer, you don't have access to its underlying allocation, so you have to allocate a byte vector and copy the bytes into it to form the underlying allocation for your `String`.

As an aside, you might want to look into the concept of "zero-copy parsing". This is essentially a way of writing parsers where all the data is stored in a single buffer, and the parsed representations of that data are represented as indices or references into parts of that buffer, rather than copying the data into new heap allocations. This can often have improved performance and reduced memory overhead compared to copying data while parsing.

I haven't watched [this talk](https://www.youtube.com/watch?v=8mA5ZwWB3M0&feature=youtu.be&t=928) but it's probably a good introduction. If you enjoy reading papers, you might like [this one](http://spw17.langsec.org/papers/chifflier-parsing-in-2017.pdf).

Reply:
"zero-copy" sounds great, and I'll learn more about this later. Currently, this sounds like a perform issue and is orthotropic to the codec framework, say, we can try the same "no-copy" approach in all the different architectures. If so, then I would postpone to study it unless I'm comfortable about the codec framework. Hopefully, we can be happy about the `enum` design.

Reply:
Yeah, that's fine --- it's definitely not necessary to rewrite everything to be zero-copy now. I only brought it up because I thought you might be interested to learn about zero-copy techniques and maybe implement them in the future.